### PR TITLE
Pitlord grouping fix/Custom ability grouping #3

### LINKED ISSUE #3: Pitlord abilities are not grouped.
Unlike all the other skills in the skill select window, his are not grouped together. This may be because he is not actually in the game and hence the ownership of those abilities to him may be bugging out.

### PR DESCRIPTION
**Pull request by TideSofDarK**
Still needs some tests. Basically what I do here is merging custom heroes txt with default heroes.txt to keep pitlord definition.

### CODE REVIEW

File: src/scripts/vscripts/pregame.lua [hunk 1]
```diff
@@ -854,7 +861,7 @@ function Pregame:networkHeroes()
 
     for heroName,heroValues in pairs(allHeroes) do
         -- Ensure it is enabled
-        if heroName ~= 'Version' and heroName ~= 'npc_dota_hero_base' and heroValues.Enabled == 1 then
+        if heroName ~= 'Version' and heroName ~= 'npc_dota_hero_base' and (heroValues.Enabled == 1 or heroName == "npc_dota_hero_abyssal_underlord") then
```
Comment (Myrl) [inline src/scripts/vscripts/pregame.lua]: Why are you checking for the name if the other file already enables Pit Lord?

**Comment by TideSofDarK** [comment]
@Myrl check out latest addition, now there is no checking for pitlord name

**Comment by Myrl** [comment]
I mean, couldn't you just have left [this](https://github.com/darklordabc/Legends-of-Dota-Redux/pull/12/commits/ac46234bd188a9ad4537a952f80b38dc3dd80707#diff-ff4e1a8ff3e8a41914865917b1b57937L857) alone? You've already set [Pitlord enabled](https://github.com/darklordabc/Legends-of-Dota-Redux/pull/12/commits/ac46234bd188a9ad4537a952f80b38dc3dd80707#diff-369a14df8f69a186cb2cc914b4b1e8bdR20767).

**Comment by TideSofDarK** [comment]
@Myrl sadly npc_heroes.txt isnt imported to scripts/npc so that line appears useless. now enabling depends on herolist.txt, where pitlord is also defined but set to 0.

**Comment by TideSofDarK** [comment]
Ok first implementation of grouping non-hero abilities is ready. There are still unknown to me things like sub-sorting which is ignored by my code. 
It still needs some tests.

**Comment by Myrl** [comment]
@TideSofDarK I figured out what was happening; the Enabled kv is not being copied. I'll make a pull request on your branch.

**Comment by TideSofDarK** [comment]
Apparently it broke stuff. Reverted for now

**Comment by TideSofDarK** [comment]
OK should be fine now. The probems was in removed "theData["Enabled"] = heroList[heroName]" line. We need this to not create pitlords hero card(because its not playable hero). Its "disabled" in herolist.txt.

File: src/panorama/scripts/custom_game/game_setup.js [hunk 2]
```diff
@@ -2686,52 +2687,52 @@ function OnSkillTabShown(tabName) {
                     if(filterInfo.shouldShow) {
                         if(useSmartGrouping) {
                             var theOwner = abilityHeroOwner[abilityName];
+                            var neutralGroup = flagDataInverse[abilityName].group;
 
-                            if(theOwner != null) {
+                            if(theOwner != null || neutralGroup != null) {
```
Comment (Myrl) [inline src/panorama/scripts/custom_game/game_setup.js]: Lines 2692 and 2694. Since theOwner and neutralGroup is only used in lines 2694, consider moving line 2694 outside and changing the condition to check for groupKey.

File: src/panorama/scripts/custom_game/game_setup.js [hunk 3]
```diff
@@ -2686,52 +2687,52 @@ function OnSkillTabShown(tabName) {
...
+                                toSort.push({
+                                    txt: groupKey,
+                                    con: groupCon,
+                                    grouped: true
+                                });
+
                                 // Making the layout much nicer
-                                if(heroBlockCounts[theOwner] == null) {
-                                    heroBlockCounts[theOwner] = 1;
+                                if(blockCounts[groupKey] == null) {
```
Comment (Myrl) [inline src/panorama/scripts/custom_game/game_setup.js]: Consider refactoring this without the null checking

File: src/panorama/scripts/custom_game/game_setup.js [hunk 4]
```diff
@@ -2686,52 +2687,52 @@ function OnSkillTabShown(tabName) {
...
+                                    blockCounts[groupKey] = 1;
                                 } else {
-                                    ++heroBlockCounts[theOwner];
+                                    ++blockCounts[groupKey];
 
-                                    if(heroBlockCounts[theOwner] == 3) {
+                                    if(blockCounts[groupKey] == 3) {
```
Comment (Myrl) [inline src/panorama/scripts/custom_game/game_setup.js]: You can probably pull this out from the if/else. Nested branching is hard to follow.

**Comment by Myrl** [comment]
Not sure if I should consider this as a bugfix or an enhancement.

**Comment by Myrl** [comment]
Thanks. Much better. We'll test it tomorrow and I'll pull it if everything works.

**Comment by darklordabc** [comment]
This is working, but I have one problem with it. It jumbles up the main, neutral and custom abilities when you have all tabs activated. I'd prefer it to prioritize hero abilities at the top, then neutral and then custom. 
image

**Comment by darklordabc** [comment]
Sorting is better. But for some reason, when you combine neutrals and normal abilities, the normal abilities get rearanged, even though neutrals are at the bottom. The order of the main abilities should not change. 
https://drive.google.com/file/d/0B7dpugx95_vaRS14Z2l6Y2U4ZEE/view

**Comment by darklordabc** [comment]
This now works correctly for neutral + normal abilities, the neutral abilities appear at the bottom of the list and it does not reorder for normal abilities. But it does not yet work for custom abilities. Custom abilities do not get placed at the bottom, below neutral abilities, as they should. 
image
I modified abilities.kv and put two neutral group abiliites as custom abilities, and this is how they showed up when I viewed normal + custom abilities.

**Comment by TideSofDarK** [comment]
So basically custom abilities that are not grouped by hero should be at bottom right?

**Comment by darklordabc** [comment]
Yes.

**Comment by darklordabc** [comment]
Wait, they should be after neutral, so at the bottom, but not the "bottom right" (there is two different interpretations to that).

**Comment by TideSofDarK** [comment]
Yeah got it.

**Comment by darklordabc** [comment]
Tested and seems to work.

**Comment by TideSofDarK** [comment]
LGTM

**Comment by Yahnich** [comment]
if my commit has been checked LGTM

**Comment by TideSofDarK** [comment]
LGTM

**Comment by darklordabc** [comment]
@Myrl what ya reckon?

**Comment by darklordabc** [comment]
This is broken for me right now. I downloaded the zip and did the usual process, and the gui is broken. Has something gone wrong with the stagging or is there a mistake in the code? 
image

**Comment by TideSofDarK** [comment]
Do u have something in console?

**Comment by darklordabc** [comment]
[ W PanoramaScript         ]: !! (s2r://panorama/scripts/custom_game/game_setup.vjs_c, line:2966, col:0) - SyntaxError: Unexpected strict mode reserved word

**Comment by TideSofDarK** [comment]
Strangely it does not have anything with my code.

**Comment by darklordabc** [comment]
Strange.

**Comment by TideSofDarK** [comment]
It breaks on not my line and also fires some strict mode error. Totally strange.

**Comment by Myrl** [comment]
It does break on your line. "private function helperSort(a,b){"

**Comment by TideSofDarK** [comment]
Oh actually its not my line but Yah's. I removed his "private"

**Comment by TideSofDarK** [comment]
Kinda strange that he decided to use "private", i've never seen usage of it in custom games.

**Comment by Yahnich** [comment]
it broke because panorama apparently only takes public functions, my bad; LGTM

**Comment by TideSofDarK** [comment]
OK i returned "return" so sorting is fine again. Again sorry for not checking Yah's PR.

**Comment by TideSofDarK** [comment]
LGTM
